lightning/src/offers/invoice.rs

```diff
@@ -0,0 +1,633 @@
+/// [module-level documentation]: self
+pub struct InvoiceBuilder<'a> {
+	bytes: &'a Vec<u8>,
+	invoice: InvoiceContents,
+}
+
+impl<'a> InvoiceBuilder<'a> {
+	pub(super) fn for_offer(
+		invoice_request: &'a InvoiceRequest, paths: Vec<BlindedPath>, payinfo: Vec<BlindedPayInfo>,
+		created_at: Duration, payment_hash: PaymentHash
+	) -> Result<Self, SemanticError> {
+		let amount_msats = match invoice_request.amount_msats() {
+			Some(amount_msats) => amount_msats,
+			None => match invoice_request.contents.offer.amount() {
+				Some(Amount::Bitcoin { amount_msats }) => {
+					amount_msats * invoice_request.quantity().unwrap_or(1)
+				},
+				Some(Amount::Currency { .. }) => return Err(SemanticError::UnsupportedCurrency),
+				None => return Err(SemanticError::MissingAmount),
+			},
+		};
+
+		let contents = InvoiceContents::ForOffer {
+			invoice_request: invoice_request.contents.clone(),
+			fields: InvoiceFields {
+				paths, payinfo, created_at, relative_expiry: None, payment_hash, amount_msats,
+				fallbacks: None, features: Bolt12InvoiceFeatures::empty(),
+				signing_pubkey: invoice_request.contents.offer.signing_pubkey(),
+			},
+		};
+
+		Self::new(&invoice_request.bytes, contents)
+	}
+
+	pub(super) fn for_refund(
+		refund: &'a Refund, paths: Vec<BlindedPath>, payinfo: Vec<BlindedPayInfo>,
+		created_at: Duration, payment_hash: PaymentHash, signing_pubkey: PublicKey
+	) -> Result<Self, SemanticError> {
+		let contents = InvoiceContents::ForRefund {
+			refund: refund.contents.clone(),
+			fields: InvoiceFields {
+				paths, payinfo, created_at, relative_expiry: None, payment_hash,
+				amount_msats: refund.amount_msats(), fallbacks: None,
+				features: Bolt12InvoiceFeatures::empty(), signing_pubkey,
+			},
+		};
+
+		Self::new(&refund.bytes, contents)
+	}
+
+	fn new(bytes: &'a Vec<u8>, contents: InvoiceContents) -> Result<Self, SemanticError> {
+		let paths = &contents.fields().paths;
+		let payinfo = &contents.fields().payinfo;
+
+		if paths.is_empty() {
+			return Err(SemanticError::MissingPaths);
+		}
+
+		if paths.len() != payinfo.len() {
+			return Err(SemanticError::InvalidPayInfo);
+		}
+
+		Ok(Self { bytes, invoice: contents })
+	}
+
+	/// Sets the [`Invoice::relative_expiry`] as seconds since [`Invoice::created_at`]. Any expiry
+	/// that has already passed is valid and can be checked for using [`Invoice::is_expired`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn relative_expiry(mut self, relative_expiry_secs: u32) -> Self {
+		let relative_expiry = Duration::from_secs(relative_expiry_secs as u64);
+		self.invoice.fields_mut().relative_expiry = Some(relative_expiry);
+		self
+	}
+
+	/// Adds a P2WSH address to [`Invoice::fallbacks`].
+	///
+	/// Successive calls to this method will add another address. Caller is responsible for not
+	/// adding duplicate addresses.
+	pub fn fallback_v0_p2wsh(mut self, script_hash: &WScriptHash) -> Self {
```

Comment:
Similar problem with returning a reference to temporary in `InvoiceFields::as_tlv_stream`. Oh, I guess we can convert it to `FallbackAddress` inside the builder. Just feels wrong to have that sort of chain of conversions and additional allocations?

Reply:
> Hmm, maybe we move `FallbackAddress` to `invoice.rs`?

Not sure I follow. It _is_ defined there.

> I guess its serialized differently because `program` is serialized with a two-byte length prefix instead of the one-byte version that is used for a bitcoin script itself (so this can't "just" be represented as a script)? Isn't that a weird encoding anyway?

There can be more than on fallback in the TLV record, thus why the encoding requires the two-byte length prefix, if I understand your question. So regardless of how it is represented, a two-byte length prefix is required per address.

Also, we aren't required to know how to interpret a version (i.e., the length of the program) at serialization time. But it can't be an arbitrary script, FWIW. Not 100% sure I'm following your point though.

Reply:
> Ah, ben pointed out its the same as for invoices, wonder if we should use the same type for that, then.
> 

BOLT 11 has a wider set of possibilities.

```
* `f` (9): `data_length` variable, depending on version. Fallback on-chain address: for Bitcoin, this starts with a 5-bit `version` and contains a witness program or P2PKH or P2SH address.
```

Looks like it uses `17` for P2PKH and `18` for P2SH, so I don't think we want to reuse the one from our BOLT 11 implementation, which is defined as:

```rust
pub enum Fallback {
        SegWitProgram {
                version: u5,
                program: Vec<u8>,
        },
        PubKeyHash([u8; 20]),
        ScriptHash([u8; 20]),
}
```

> Right, I think my point here is it seemed strange that the BOLT12 structs had a five methods to set script stuff, which can be reused on BOLT11 and shutdown, and seemed incredibly weird to expose in BOLT12 at all.

Three methods, not five. And they are in terms of `rust-bitcoin` types, which is the norm in that crate.

We can't use `ShutdownScript` as-is since internally it is represented as a `Script` which can't easily be used with `InvoiceTlvStreamRef`.

> Its not really worth this long a discussion around it, though.

Agreed. We aren't introducing any public types here. We can always change it later.